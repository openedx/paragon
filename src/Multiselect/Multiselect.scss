@import "variables";

.pgn__multiselect {
  .pgn__multiselect__floating-label {
    color: $gray-500;
    background: transparent;
    font-size: $font-size-base;
    user-select: none;
    pointer-events: none;
    position: absolute;
    transform-origin: left top;
    transition: all .15s ease-in-out;
    top: calc(50% - $multiselect-placeholder-height / 2);
    left: $multiselect-padding-x;
    line-height: $multiselect-placeholder-height;
    grid-area: auto;
    max-width: 100%;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;

    &.show {
      z-index: $multiselect-floating-label-zindex;
      padding: 0 .5rem;
      background: $white;
      line-height: normal;
      font-size: $font-size-sm;
      left: $multiselect-padding-x / 2;
      top: -1rem;
    }
  }

  &.dark {
    .pgn__multiselect__floating-label {
      &.show {
        padding: 0;
        color: $light-700;
        background: transparent;
        top: -1.25rem;
        left: 0;
      }
    }

    .pgn__multiselect__chip {
      background: $dark-200;

      * {
        color: $white;
      }
    }
  }

  &.disabled {
    opacity: .3;
    pointer-events: none;
  }

  &.error {
    .pgn__multiselect__control {
      border-color: $danger-300;
      box-shadow: 0 0 0 1px $danger-300;
    }
  }
}

div.pgn__multiselect__control {
  border-radius: 0;
  border-color: $gray-500;
  box-shadow: none;

  & > div {
    padding-right: $multiselect-padding-x / 2;
  }

  & > div:nth-child(2) {
    padding-left: $multiselect-padding-x;
  }

  &:hover {
    border-color: $black;
  }

  &.is-focus {
    box-shadow: 0 0 0 1px $primary-500;
  }
}

.pgn__multiselect__error-message {
  color: $brand-500;
  font-size: $font-size-sm;
}

div.pgn__multiselect__menu {
  z-index: $multiselect-menu-zindex;
  margin-top: $multiselect-menu-offset-top;

  .is-focus {
    background: $light-300;
  }
}

div.pgn__multiselect__chip {
  background: $light-500;
  border-radius: 4px;
}

div.pgn__multiselect__chip-label {
  font-size: $font-size-sm;
  padding: 2px 2px 2px 8px;
}

div.pgn__multiselect__chip-remove {
  border-radius: 0 4px 4px 0;

  .pgn__icon {
    width: 1.25rem;
    height: 1.25rem;
  }

  &:hover,
  &:focus {
    color: $white;
    background: $black;
  }
}

.pgn__multiselect__indicator {
  width: auto;
  height: auto;
  padding: 8px;

  &:hover {
    fill: $white;
  }

  &:focus {
    fill: $primary-500;
  }
}

div.pgn__multiselect__input {
  padding-top: 0;
  padding-bottom: 0;
}
