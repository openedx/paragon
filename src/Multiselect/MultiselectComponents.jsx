import React from 'react';
import classNames from 'classnames';
import { components } from 'react-select';
import PropTypes from 'prop-types';
import Icon from '../Icon';
import Chip from '../Chip';
import { Close, KeyboardArrowDown } from '../../icons';
import IconButton from '../IconButton';

const ClearIndicator = ({ isDisabled, innerProps }) => (
  <IconButton
    src={Close}
    iconAs={Icon}
    alt="Clear"
    variant="primary"
    className="pgn__multiselect__indicator"
    tabIndex={isDisabled ? -1 : 0}
    onKeyPress={innerProps.onMouseDown}
    {...innerProps}
  />
);

const DropdownIndicator = ({ isDisabled, innerProps }) => (
  <IconButton
    src={KeyboardArrowDown}
    iconAs={Icon}
    alt="Open"
    variant="primary"
    className="pgn__multiselect__indicator"
    tabIndex={isDisabled ? -1 : 0}
    onKeyPress={innerProps.onMouseDown}
    {...innerProps}
  />
);

const Input = (props) => (
  <components.Input className="pgn__multiselect__input" {...props} />
);

const ValueContainer = ({ children, ...props }) => (
  <components.ValueContainer {...props}>
    {React.Children.map(children, (child) => (child && child.type !== components.Placeholder ? child : null))}
  </components.ValueContainer>
);

const Control = ({
  isFocused,
  hasValue,
  children,
  selectProps,
  ...props
}) => (
  <>
    <components.Control
      className={classNames('pgn__multiselect__control', { 'is-focus': isFocused })}
      {...props}
      hasValue={hasValue}
      isFocused={isFocused}
    >
      <components.Placeholder
        className={classNames('pgn__multiselect__floating-label',
          { show: hasValue || selectProps.inputValue || isFocused })}
        {...props}
      >
        {selectProps.placeholder}
      </components.Placeholder>
      {React.Children.map(children, (child) => React.cloneElement(child, { ...child.props, isFocused }))}
    </components.Control>
  </>
);

const Menu = (props) => (
  <components.Menu className="pgn__multiselect__menu" {...props} />
);

const Option = (props) => (
  <components.Option className={classNames({ 'is-focus': props.isFocused })} {...props} />
);

const MultiValueContainer = ({ variant, children }) => {
  const chipLabel = children[0]?.props?.children;
  const { onClick = () => {} } = children[1]?.props?.innerProps || {};
  return (
    <Chip
      variant={variant}
      iconAfter={Close}
      onIconAfterClick={onClick}
    >
      {chipLabel}
    </Chip>
  );
};

const indicatorPropTypes = {
  isDisabled: PropTypes.bool,
  innerProps: PropTypes.shape({
    onMouseDown: PropTypes.func.isRequired,
  }).isRequired,
};

ClearIndicator.propTypes = indicatorPropTypes;
DropdownIndicator.propTypes = indicatorPropTypes;

ClearIndicator.defaultProps = {
  isDisabled: false,
};

DropdownIndicator.defaultProps = {
  isDisabled: false,
};

ValueContainer.propTypes = {
  children: PropTypes.node.isRequired,
};

Control.propTypes = {
  isFocused: PropTypes.bool.isRequired,
  hasValue: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  selectProps: PropTypes.shape({
    placeholder: PropTypes.string,
    inputValue: PropTypes.string,
  }).isRequired,
};

Option.propTypes = {
  isFocused: PropTypes.bool.isRequired,
};

MultiValueContainer.propTypes = {
  children: PropTypes.node.isRequired,
  variant: PropTypes.oneOf(['light', 'dark']),
};

MultiValueContainer.defaultProps = {
  variant: 'light',
};

export const multiselectComponents = {
  ClearIndicator,
  DropdownIndicator,
  ValueContainer,
  Control,
  Menu,
  Option,
  MultiValueContainer,
  Input,
};
export default multiselectComponents;
