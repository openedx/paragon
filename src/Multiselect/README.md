---
title: 'Multiselect'
type: 'component'
components:
- Multiselect
categories:
- Forms
status: 'New'
designStatus: 'Done'
devStatus: 'Done'
---

A text input that allows the user to select multiple options. `Multiselect` is a modified version of React Select. For more information and props, please, check [React Select](https://react-select.com/home) documentation.

## Basic Usage

```jsx live
() => {
  const colourOptions = [
    {
      value: 'ocean',
      label: 'Ocean',
      color: '#00B8D9',
      isFixed: true
    },
    {
      value: 'blue',
      label: 'Blue',
      color: '#0052CC',
      isDisabled: true
    },
    {
      value: 'purple',
      label: 'Purple',
      color: '#5243AA'
    },
    {
      value: 'red',
      label: 'Red',
      color: '#FF5630',
      isFixed: true
    },
    {
      value: 'orange',
      label: 'Orange',
      color: '#FF8B00'
    },
    {
      value: 'yellow',
      label: 'Yellow',
      color: '#FFC400'
    },
    {
      value: 'green',
      label: 'Green',
      color: '#36B37E'
    },
    {
      value: 'forest',
      label: 'Forest',
      color: '#00875A'
    },
    {
      value: 'slate',
      label: 'Slate',
      color: '#253858'
    },
    {
      value: 'silver',
      label: 'Silver',
      color: '#666666'
    },
  ];
  
  const handleChange = (e) => console.log(e);

  return (
    <Stack direction="vertical">
      <div className="py-4">
        <Multiselect
          onChange={handleChange}
          placeholder="Select colour"
          options={colourOptions}
        />
      </div>
      <div className="py-4">
        <Multiselect
          disabled
          onChange={handleChange}
          placeholder="Select colour"
          options={colourOptions}
        />
      </div>
      <div className="bg-dark p-4">
        <Multiselect
          onChange={handleChange}
          variant="dark"
          placeholder="Select colour"
          options={colourOptions}
        />
      </div>
    </Stack>
  );
};
```

### Error state

```jsx live
() => {
  const colourOptions = [
    {
      value: 'ocean',
      label: 'Ocean',
      color: '#00B8D9',
    },
    {
      value: 'green',
      label: 'Green',
      color: '#36B37E'
    },
    {
      value: 'forest',
      label: 'Forest',
      color: '#00875A'
    },
  ];
  
  const handleChange = (e) => console.log(e);

  return (
    <Multiselect
      error="There is an error!"
      onChange={handleChange}
      placeholder="Select colour"
      options={colourOptions}
    />
  );
};
```
