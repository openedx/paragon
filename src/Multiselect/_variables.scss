$multiselect-placeholder-height:     1.5rem;
$multiselect-padding-x:              1rem;
$multiselect-floating-label-zindex:  1;
$multiselect-menu-zindex:            2;
$multiselect-menu-offset-top:        2px;
