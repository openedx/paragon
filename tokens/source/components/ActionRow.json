{
  "action-row": {
    "gap": {
      "x": { "value": ".5rem", "themeable": true, "source": "$action-row-gap-x" },
      "y": { "value": ".5rem", "themeable": true, "source": "$action-row-gap-y" }
    }
  }
}
