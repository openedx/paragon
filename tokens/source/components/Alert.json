{
  "alert": {
    "padding": {
      "y": { "value": "1.5rem", "themeable": true, "source": "$alert-padding-y" },
      "x": { "value": "1.5rem", "themeable": true, "source": "$alert-padding-x" }
    },
    "margin": {
      "bottom": { "value": "1rem", "themeable": true, "source": "$alert-margin-bottom" }
    },
    "border": {
      "radius": { "value": "{border.radius.base}", "themeable": true, "source": "$alert-border-radius" },
      "width": { "value": "0", "themeable": true, "source": "$alert-border-width" }
    },
    "font": {
      "weight": {
        "link": { "value": "{font.weight.normal.value}", "themeable": true, "source": "$alert-link-font-weight" }
      },
      "size": { "value": ".875rem", "themeable": true, "source": "$alert-font-size" }
    },
    "color": {
      "title": { "value": "{color.black.value}", "themeable": true, "source": "$alert-title-color" },
      "content": { "value": "{color.gray.700.value}", "themeable": true, "source": "$alert-content-color" }
    },
    "box-shadow": {
      "value": "0 1px 2px rgba(0, 0, 0, .15), 0 1px 4px rgba(0, 0, 0, .15)", 
      "themeable": true, 
      "source": "$alert-box-shadow"
    },
    "level": {
      "bg": { "value": "-10", "themeable": true, "source": "$alert-bg-level" },
      "border": { "value": "-9", "themeable": true, "source": "$alert-border-level" },
      "color": { "value": "6", "themeable": true, "source": "$alert-color-level" }
    },
    "icon": {
      "space": { "value": ".8rem", "themeable": true, "source": "$alert-icon-space" }
    },
    "line-height": { "value": "1.5rem", "themeable": true, "source": "$alert-line-height" },
    "actions": {
      "gap": { "value": "{spacing.base.3.value}", "themeable": true, "source": "$alert-actions-gap" }
    }
  }
}
