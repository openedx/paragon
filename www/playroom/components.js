const components = require('../../src');

// Some icons have the same names as Paragon's components,
// so we rename them here first before exporting to playroom
const {
  Badge: BadgeIcon,
  CheckBox: CheckBoxIcon,
  Image: ImageIcon,
  Input: InputIcon,
  Menu: MenuIcon,
  Spinner: SpinnerIcon,
  Tab: TabIcon,
  ...restIcons
} = require('../../icons');

module.exports = {
  BadgeIcon,
  CheckBoxIcon,
  InputIcon,
  ImageIcon,
  MenuIcon,
  SpinnerIcon,
  TabIcon,
  ...restIcons,
  ...components,
};
