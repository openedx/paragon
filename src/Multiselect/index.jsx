import React from 'react';
import Select from 'react-select';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import components, { MultiValueContainer } from './MultiselectComponents';

const Multiselect = React.forwardRef(({
  className,
  variant,
  disabled,
  error,
  ...props
}, ref) => (
  <>
    <Select
      isMulti
      isDisabled={disabled}
      ref={ref}
      components={{
        ...components,
        MultiValueContainer: (containerProps) => <MultiValueContainer variant={variant} {...containerProps} />,
      }}
      className={classNames('pgn__multiselect', className, { dark: variant === 'dark', disabled, error })}
      {...props}
    />
    {error && <div className="pgn__multiselect__error-message">{error}</div>}
  </>
));

Multiselect.defaultProps = {
  className: undefined,
  variant: 'light',
  disabled: false,
  error: undefined,
};

Multiselect.propTypes = {
  /** A class name to append to the base element. */
  className: PropTypes.string,
  /** The `Multiselect` style variant to use. */
  variant: PropTypes.oneOf(['light', 'dark']),
  /** Specifies whether the `Multiselect` is disabled. */
  disabled: PropTypes.bool,
  /** Applies the error styles and displays the error text if it is specified. */
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
};

export default Multiselect;
