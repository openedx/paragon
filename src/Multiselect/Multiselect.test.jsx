import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Multiselect from './index';
import components from './MultiselectComponents';

// eslint-disable-next-line react/prop-types
const TestComponent = (props) => <div>{props.children}</div>;

describe('<Multiselect />', () => {
  it('successfully renders', () => {
    const tree = renderer.create(<Multiselect />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders with correct class when className is passed', () => {
    const testClassName = 'classname';
    const wrapper = mount(<Multiselect className={testClassName} />);
    const multiselect = wrapper.find('.pgn__multiselect').at(0);
    expect(multiselect.hasClass(testClassName)).toEqual(true);
  });
  it('renders with the dark variant', () => {
    const wrapper = mount(<Multiselect variant="dark" />);
    const multiselect = wrapper.find('.pgn__multiselect').at(0);
    expect(multiselect.hasClass('dark')).toEqual(true);
  });
  it('renders with disabled prop', () => {
    const wrapper = mount(<Multiselect disabled />);
    const multiselect = wrapper.find('.pgn__multiselect').at(0);
    expect(multiselect.hasClass('disabled')).toEqual(true);
  });
  it('renders with error prop', () => {
    const errorText = 'error text';
    const wrapper = mount(<Multiselect error={errorText} />);
    const multiselect = wrapper.find('.pgn__multiselect').at(0);
    expect(multiselect.hasClass('error')).toEqual(true);
    expect(wrapper.find('.pgn__multiselect__error-message').text()).toEqual(errorText);
  });

  describe('Multiselect Components', () => {
    it('Menu has base class', () => {
      const wrapper = mount(<components.Menu cx={() => {}} getStyles={() => {}} />);
      expect(wrapper.find('.pgn__multiselect__menu').length).toBeTruthy();
    });
    it('Option gets is-focus class on hover', () => {
      const wrapper = mount(<components.Option cx={() => {}} getStyles={() => {}} isFocused />);
      expect(wrapper.find('.is-focus').length).toBeTruthy();
    });
    it('MultiValueContainer has base class', () => {
      const wrapper = mount((
        <components.MultiValueContainer>
          <TestComponent innerProps={{}}>1</TestComponent>
          <TestComponent innerProps={{}}>2</TestComponent>
        </components.MultiValueContainer>
      ));
      expect(wrapper.find('.pgn__chip').length).toBeTruthy();
    });
    it('ClearIndicator has base class', () => {
      const wrapper = mount(<components.ClearIndicator innerProps={{ onMouseDown: () => {} }} />);
      expect(wrapper.find('.pgn__multiselect__indicator').length).toBeTruthy();
    });
  });
});
