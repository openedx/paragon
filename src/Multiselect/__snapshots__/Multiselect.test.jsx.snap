// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Multiselect /> successfully renders 1`] = `
<div
  className="pgn__multiselect css-b62m3t-container"
  onKeyDown={[Function]}
>
  <span
    className="css-1f43avz-a11yText-A11yText"
    id="react-select-2-live-region"
  />
  <span
    aria-atomic="false"
    aria-live="polite"
    aria-relevant="additions text"
    className="css-1f43avz-a11yText-A11yText"
  />
  <div
    className="pgn__multiselect__control css-1s2u09g-control"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="pgn__multiselect__floating-label css-14el2xx-placeholder"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      Select...
    </div>
    <div
      className=" css-319lph-ValueContainer"
    >
      <div
        className="pgn__multiselect__input css-6j8wv5-Input"
        data-value=""
      >
        <input
          aria-autocomplete="list"
          aria-describedby="react-select-2-placeholder"
          aria-expanded={false}
          aria-haspopup={true}
          autoCapitalize="none"
          autoComplete="off"
          autoCorrect="off"
          className=""
          disabled={false}
          id="react-select-2-input"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          role="combobox"
          spellCheck="false"
          style={
            Object {
              "background": 0,
              "border": 0,
              "color": "inherit",
              "font": "inherit",
              "gridArea": "1 / 2",
              "label": "input",
              "margin": 0,
              "minWidth": "2px",
              "opacity": 1,
              "outline": 0,
              "padding": 0,
              "width": "100%",
            }
          }
          tabIndex={0}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className=" css-1hb7zxy-IndicatorsContainer"
    >
      <span
        className=" css-1okebmr-indicatorSeparator"
      />
      <button
        aria-hidden="true"
        aria-label="Open"
        className="btn-icon btn-icon-primary btn-icon-md pgn__multiselect__indicator"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="btn-icon__icon-container"
        >
          <span
            className="pgn__icon btn-icon__icon"
          >
            <svg
              aria-hidden={true}
              focusable={false}
              height={24}
              role="img"
              viewBox="0 0 24 24"
              width={24}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
              />
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
